### ISSUE TITLE
SDK bundle: no way to pass different toolset paths based on host supportedTriples?

### ISSUE THREAD

**Issue by finagolfin** [swiftlang/swift-package-manager#8029]
### Is it reproducible with SwiftPM command-line tools: `swift build`, `swift test`, `swift package` etc?

- [X] Confirmed reproduction steps with SwiftPM CLI. The description text _must_ include reproduction steps with either of command-line SwiftPM commands, `swift build`, `swift test`, `swift package` etc.

### Description

[The SDK Bundle spec provides a way to specify the host triples supported in `info.json`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md#swift-sdk-bundles), but then doesn't seem to use that info anywhere.

I thought one obvious reason to specify those was to provide toolset executables for multiple hosts, but there doesn't appear to be any way to specify the different hosts in [`swift-sdk.json`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md#swift-sdkjson-files) or [the toolset files](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md#toolsetjson-files).

For example, if I want to specify a different `cCompiler` in the toolset for linux x86_64, I can configure that. However, if the same SDK bundle is used on linux AArch64 or macOS x86_64, there doesn't appear to be a way to specify alternate paths for those executables also.

@MaxDesiatov, an oversight in the SDK bundle spec?

### Expected behavior

The doc to provide a way to specify multiple host-platform-specific executables in an SDK bundle

### Actual behavior

I see no info on this.

### Steps to reproduce

Not implemented

### Swift Package Manager version/commit hash

All

### Swift & OS version (output of `swift --version ; uname -a`)

All

**Comment by finagolfin** [comment]
@MaxDesiatov, any plans on what to do here? Presumably you'd have to provide a way to inject different host executable paths into the toolsets, and I was just thinking that for support of external C/C++ platform SDKs to ever work well, after the bugs reported in #8958 are worked out, we'd also need to provide a way to inject those paths into the toolsets, so that path-specific flags like `-I /(sdkRootPath)/opt/local/include/` could also be added to the toolsets.

We could probably do these path variable injections all at once, wondering what your plan is for this host executable one.

**Comment by dschaefer2** [comment]
@owenv Thoughts on this?

**Comment by dschaefer2** [comment]
As I try to understand this more, the host triple appears to just be used to filter which SwiftSDK in the artifact bundle is to be used on the given host. Is that the issue here, that it doesn't filter? That would be more a SwiftPM issue since I don't think SwiftBuild looks at the info.json file for the artifact bundle. Or does it Owen?

**Comment by finagolfin** [comment]
Not a SwiftBuild issue per se- I don't know where this would be implemented in SwiftPM after the swift-build transition- just that the linked SDK bundle spec provides for ways to include executables for each host, but that multi-OS part of the proposal appears to be unimplemented in SwiftPM.

Since this would likely be implemented by passing a path in as a variable to a toolset, it would be much more broadly useful than just filling that multi-OS executable gap, including being able to pass external paths to compiler flags and so on.

**Comment by dschaefer2** [comment]
My take is that it selects whole SwiftSDKs based on host, not just toolchain files. So it should be a matter of SwiftPM instantiating the right SwiftSDK and it's toolset.json would get sent in to the build. Or am I mistaken

**Comment by dschaefer2** [comment]
And looking at the code, I see:
```
                    guard variant.isSupporting(hostTriple: hostTriple) else { continue }
```
in the selectSwiftSDK function in SwiftSDKBundle.

**Comment by finagolfin** [comment]
Sure, it already has a way to choose an SDK based on the host too, but I don't see any way to associate the SDK with a single host OS triple either. Anyway, the only reason that association would be useful is to provide different versions of a tool in a bundle for each host OS.

Basically, the info.json provides a way to list supported host OS triples and the toolset definition says you can ship tools that run on the host, but the proposal language on where you look for those multi-host tools and a SwiftPM implementation is missing!

Maybe I missed it somewhere: try building a simple SDK bundle that provides multi-OS versions of a tool to run on multiple hosts if you don't believe me.

**Comment by dschaefer2** [comment]
I am just replying to the first line of your description. The info.json file is part of the artifact bundle spec, not the SwiftSDK spec. From my read of the code, it's doing what it promised, selecting the SwiftSDK artifact based the supportedTriples field checked against the host triple (i.e. that's what isSupporting() is doing, see below). And of course the toolchain definition for the host tools is part of the SwiftSDK that gets selected.

```
extension SwiftSDKBundle.Variant {
    /// Whether the given host triple is supported by this SDK variant
    internal func isSupporting(hostTriple: Triple) -> Bool {
        guard let supportedTriples = metadata.supportedTriples else {
            // No supportedTriples means the SDK can be universally usable
            return true
        }
        return supportedTriples.contains(where: { variantTriple in
            hostTriple.isRuntimeCompatible(with: variantTriple)
        })
    }
}
```

I recommend you set a breakpoint there and if it's not doing that, we can focus on a bug fix.

**Comment by dschaefer2** [comment]
It feels like you're asking for a change to the Swift SDK spec to move the host triples down there. That's a bigger deal that would require an evolution proposal.

Which I wouldn't mind since I'm working hard to replace other uses of artifact bundles. Swift SDKs in my future vision would start at the swift-sdk.json file.

**Comment by finagolfin** [comment]
I think you just diagnosed the problem: that `isSupporting(hostTriple:)` method is useless because all it does is check that the artifact bundle lists a host triple that matches the host that it is being installed in, or the SDK bundle's `info.json` lists no host triples and thus is considered "universal." Since as you say, the individual SDKs have no way to be associated with a single host, we cannot use the SDK's toolset to list a path to separate executable tools per host.

How precisely we fix this, ie whether it requires updating the proposal in addition to SwiftPM, is immaterial to me. All I care about is that [the language in the toolset section about being able to use local tools](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0387-cross-compilation-destinations.md#toolsetjson-files) actually gets implemented for cross-OS scenarios, eg we can customize a ld.lld tool in the bundle's toolset, then specify local paths to that executable for each of the host triples.

Currently, that cross-OS tool bundling is described as possible, but the specific mechanism to include tools in the SDK bundle for multiple host triples doesn't exist.

**Comment by dschaefer2** [comment]
My point is that it does exist in the current spec, it's just pretty awkward and involves bundling multiple SDKs into the same artifact bundle, one per host, and then set up a directory structure in the bundle that lets you share pieces between the hosts.

I'm open to proposals and PRs that make that easier to do in a single SDK since that aligns with the long term direction I have for Swift SDKs.

**Comment by finagolfin** [comment]
> My point is that it does exist in the current spec, it's just pretty awkward and involves bundling multiple SDKs into the same artifact bundle, one per host, and then set up a directory structure in the bundle that lets you share pieces between the hosts.

I see, yeah, that may be possible, I'll check. My point is that as soon as the SDK is registered as supporting multiple hosts or universal, then there is no longer a mechanism to supply a path to a tool for each of the supported host triples. I'm pretty sure that was the intent of supplying multiple host triples in the first place, ie to be able to supply multiple versions of custom tools for each host.

**Comment by dschaefer2** [comment]
Correct. The semantics of supported triples is that the contents of the artifact supports any of the listed hosts. If you have a tool in there that only supports one, then you need to create a different artifact for it that only lists that host.

**Comment by finagolfin** [comment]
No, that was not the semantics as designed. It would be pointless to say a SDK supports multiple host triples, then say you have to separate each SDK as only supporting one host triple for it to work. If you are simply saying that is the only way to make it work now as it has been sloppily implemented, sure, your hack may work.

I'm saying we need to fix it so the semantics of supporting multiple host triples actually has any meaning, as that is currently broken and useless.

**Comment by dschaefer2** [comment]
No argument there. I'm currently working on a replacement for artifact bundles anyway. These things need to be in the package manifest. Swift SDKs are going to need to find a new home at the end of the day.
